### PR TITLE
Cleaning ct after enabling policy

### PR DESCRIPTION
**Pull request by aanm**
**Summary of changes**:

```release-note
Enforcing policy after loading policy when endpoints where in "default" policy enforcement mode.
```

There's still one runtime tests missing as written on the runtime tests file but the review can be done now.

### CODE REVIEW

**Comment by tgraf** [comment]
rebased

**Comment by tgraf** [comment]
Logic looks OK, we can eventually get fancy and only clear CT entries in one direction to account for  individual ingress/egress policy changes but this is OK to get started.

File: contrib/vagrant/start.sh [hunk 1]
```diff
@@ -358,6 +358,7 @@ else
     if test ${CILIUM_USE_ENVOY}; then
         echo 'CILIUM_USE_ENVOY=1' >> /etc/sysconfig/cilium
     fi
+    echo 'CILIUM_USE_ENVOY=1' >> /etc/profile
```
Comment (tgraf) [inline contrib/vagrant/start.sh]: This seems seems

**Review by tgraf — CHANGES_REQUESTED** [review]

File: daemon/daemon.go [hunk 2]
```diff
@@ -322,6 +322,15 @@ func (d *Daemon) CleanCTEntries(e *endpoint.Endpoint, isCTLocal bool, ips []net.
 	endpointmanager.RmCTEntriesOf(!d.conf.IPv4Disabled, e, isCTLocal, ips, idsToRm)
 }
 
+// FlushCTEntries flushes the connection tracking of the given endpoint
+// where the given endpoint IPs' match the CT entry fields.
+// isCTLocal should bet set as true if the endpoint's CT table is either
+// local or not (if is not local then is assumed to be global).
+// Implementation of pkg/endpoint.Owner interface
```
Comment (ianvernon) [inline daemon/daemon.go:329]: Thanks for adding this comment - makes tracking interface implementation easier!

File: pkg/endpoint/owner.go [hunk 3]
```diff
@@ -83,6 +83,12 @@ type Owner interface {
 	// isCTLocal should bet set as true if the endpoint's CT table is either
 	// local or not (if is not local then is assumed to be global).
 	CleanCTEntries(e *Endpoint, isCTLocal bool, ips []net.IP, idsToRm policy.RuleContexts)
+
+	// FlushCTEntries flushes the connection tracking of the given endpoint
+	// where the given endpoint IPs' and the idsToKeep don't match any of the CT entry fields.
+	// isCTLocal should bet set as true if the endpoint's CT table is either
```
Comment (ianvernon) [inline pkg/endpoint/owner.go]: nit: bet --> set .

File: pkg/endpoint/policy.go [hunk 4]
```diff
@@ -462,7 +470,7 @@ func (e *Endpoint) regenerateL3Policy(owner Owner, repo *policy.Repository, revi
 // fails for other endpoints.
 //
 // Must be called with endpoint mutex held.
-func (e *Endpoint) regeneratePolicy(owner Owner, opts models.ConfigurationMap) (bool, policy.RuleContexts, error) {
+func (e *Endpoint) regeneratePolicy(owner Owner, opts models.ConfigurationMap) (bool, bool, policy.RuleContexts, policy.RuleContexts, error) {
```
Comment (ianvernon) [inline pkg/endpoint/policy.go]: Given the many return values of this function, document what the second boolean means and what the policy.RuleContexts represent .

File: tests/21-ct-clean-up.sh [hunk 5]
```diff
@@ -73,7 +87,7 @@ function test_reachability(){
   local container="${1}"
   local dst_ip="${2}"
   local http_path="${3}"
-  log "trying to curl http://${dst_ip}:80${http_path} from client container (should work)"
+  log "trying to docker exec -i "${container}" bash -c \"curl -s -o /dev/null -w \"%{http_code}\" --connect-timeout $TIMEOUT -XGET http://${dst_ip}:80${http_path}\" from client container (should work)"
```
Comment (ianvernon) [inline tests/21-ct-clean-up.sh]: I think you can remove the quotes before and after `${container}`

File: tests/21-ct-clean-up.sh [hunk 6]
```diff
@@ -107,6 +121,21 @@ function test_unreachability(){
...
 
+function test_reachability_port(){
+  local DROP_TIMEOUT="2"
+
+  local container="${1}"
+  local dst_ip="${2}"
+  local dst_port="${3}"
+  local http_path="${4}"
+  log "trying to curl http://${dst_ip}:${dst_port}${http_path} from client container (shouldn't work)"
+  log "docker exec -d "${container}" bash -c \"curl -s -o /dev/null -w \"%{http_code}\" --connect-timeout $DROP_TIMEOUT -XGET http://${dst_ip}:${dst_port}${http_path}\""
```
Comment (ianvernon) [inline tests/21-ct-clean-up.sh]: I think you can remove the quotes before and after `${container}`

File: tests/21-ct-clean-up.sh [hunk 7]
```diff
@@ -107,6 +121,21 @@ function test_unreachability(){
...
+  ret="$(docker exec -d "${container}" bash -c "curl -s -o /dev/null -w \"%{http_code}\" --connect-timeout $DROP_TIMEOUT -XGET http://${dst_ip}:${dst_port}${http_path}")"
```
Comment (ianvernon) [inline tests/21-ct-clean-up.sh]: I think you can remove the quotes before and after `${container}`

File: tests/21-ct-clean-up.sh [hunk 8]
```diff
@@ -452,4 +487,147 @@ test_unreachability "client" "$SERVER_3_IP4" "/public"
...
+
+  log "checking if only port 81 and port 82 are open the ct table since they are specified on a imported rule and are not going through the proxy"
+
+  bef_client_server_4_ct_entries=$(count_ct_entries_of "\[${SERVER_4_IP4}\]" "\[${CLIENT_IP}\]" "${CLIENT_SEC_ID}")
+  bef_client4_server_4_ct_entries=$(count_ct_entries_of "${SERVER_4_IP4}" "${CLIENT_IP4}" "${CLIENT_SEC_ID}")
+
+  check_ct_entries_of_le "${bef_client_server_4_ct_entries}" 2 "${SERVER_4_IP}" "${CLIENT_IP}"
+  check_ct_entries_of_le "${bef_client4_server_4_ct_entries}" 2 "${SERVER_4_IP4}" "${CLIENT_IP4}"
+else
+  log "Not using envoy proxy: skipping rest of CT tests due oxy bug"
```
Comment (ianvernon) [inline tests/21-ct-clean-up.sh]: due --> due to.

What oxy bug? Please specify and ref: a GH issue.

**Review by ianvernon — APPROVED** [review]
Nothing blocking, looks good.

File: tests/21-ct-clean-up.sh (hunk 8 above)
Comment (aanm) [reply tests/21-ct-clean-up.sh]: Done #2240

File: tests/21-ct-clean-up.sh (hunk 7 above)
Comment (aanm) [reply tests/21-ct-clean-up.sh]: Done

File: tests/21-ct-clean-up.sh (hunk 6 above)
Comment (aanm) [reply tests/21-ct-clean-up.sh]: Leftover

File: tests/21-ct-clean-up.sh (hunk 5 above)
Comment (aanm) [reply tests/21-ct-clean-up.sh]: Done

File: pkg/endpoint/policy.go (hunk 4 above)
Comment (aanm) [reply pkg/endpoint/policy.go]: Done

File: pkg/endpoint/owner.go (hunk 3 above)
Comment (aanm) [reply pkg/endpoint/owner.go]: Done

**Review by manalibhutiyani — APPROVED** [review]
looks good

**Review by tgraf — APPROVED** [review]

**Pull request merged**